woocommerce/woocommerce-ios
# [Bundle products in order form] Order form: fix product selector view model being initialized more than once

Closes: #11198 


## Why


While inspecting the API requests debugging the variation issues https://github.com/woocommerce/woocommerce-ios/pull/11186#pullrequestreview-1736581081, I noticed that `ProductSelectorViewModel` was initialized every time a state in the product selector changes - selecting/deselecting a product, configuring a bundle product, etc. This is likely due to how SwiftUI works, the view is refreshed whenever a state (`ProductSelectorViewModel`) that is passed to the view (`ProductSelectorNavigationView`) changes. This behavior of reinitializing the view model every time a state changes results in same products API requests being made throughout the product selector session. In this PR, it fixes the duplicate API requests by not creating a product selector view model in the SwiftUI view.

## How

A straightforward fix would be to make `ProductSelectorViewModel` an optional observable property, and use the nullable state for the sheet in `OrderForm` through `.sheet(item:onDismiss:content:)` instead of the `isPresented` boolean state. This was my original implementation, but I noticed the dismissal stopped working. This is because the `isPresented` boolean state is passed to `ProductSelectorView` in order to dismiss the product selector in addition to the CTA from the order form to show the product selector. Without refactoring too much, I decided to keep the `isPresented` boolean state but move it from `OrderForm.ProductsSection.showAddProduct` to `ProductSelectorViewModel.isProductSelectorPresented` as the source of truth for whether the product selector is presented. A new function `EditableOrderViewModel.toggleProductSelectorVisibility` was added and is triggered when a merchant taps to view the product selector. When toggling the product selector to be visible, `ProductSelectorViewModel` is initialized and used to show the content of the presentation sheet in `OrderForm`.

Regarding the change to add `ProductSelectorViewModel.addSelection` function: this is to support external product selection like from configuring a bundle product so that the `# Products selected` text for the CTA at the bottom of the product selector can be updated. `ProductSelectorViewModel` is internally tracking the number of selected items (products and variations) to be shown in the CTA. However, product selection (from bundle configuration and maybe other extensions in the future) can happen outside of the core logic in `ProductSelectorViewModel` and multiple of the same product can be selected (right now `ProductSelectorViewModel` expects a product to be either selected or deselected). It'd be great if we could refactor the source of truth of selected items so that the same info in `EditableOrderViewModel.selectedProducts` and `ProductSelectorViewModel.selectedItemsIDs` can be consolidated.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items.

- Go to the orders tab
- Tap `+` to create an order
- Enable a tool to observe network requests
- Tap `Add Products` --> notice that there are 2 products API requests - this is a pre-existing behavior where both `onLoadTrigger` and `synchronizeProductFilterSearch` are both triggered to load products from initialization. This is beyond the scope of this PR, and I'll confirm this behavior separately
- Tap to select and deselect at least one product --> notice that there should be no extra products API requests except for loading product variations from tapping a variable product. The CTA text at the bottom should be updated accordingly
- Tap on a bundle product in the prerequisite to configure the bundle
- Update the configuration if needed so that the configuration is valid, then tap `Done` to save the bundle --> the CTA text at the bottom should be updated accordingly

---
